magento/pwa-studio
# Uses graphQl mutation to create/fetch cartId.

## Description

NOTE: REST endpoint returns data for "authed" cart created through graphql. To repro, log in, add item to cart, then delete sign in token from local storage and refresh. Cart details should be fetched but you should not be able to add/update/remove. If you take an action on the cart it will reset to a new cart. This could be a bug in REST. It should not matter since those actions are moving to graphql which won't return data for unauthed user.

This PR precedes #1987.

Changes include:

* **Proper use of cart id for quote id**

GraphQL endpoints use the masked (hashed) cartId regardless of auth state. REST uses the "actual" (table key value) cartId as quoteId for authed requests and masked cartId as quoteId for guest requests. This means that while we are in this intermediate state of _some_ requests being through graphql (such as `createCart` and `addItemToCart`) we have to make sure the REST calls use the appropriate cartId.

* **Passing of fetch/invoker to `createCart`**

The majority of this PR is to pass the mutation callback into the actions that may need to call `createCart`.

* **Cleanup of reducer state**

We were passing and spreading unnecessary info to the store in the `receive` actions of certain asyncActions. I removed this as it was extraneous.

## Related Issue





Closes [PWA-140](https://jira.corp.magento.com/browse/PWA-140).

## Acceptance 



### Verification Stakeholders

@jimbo

### Specification



### Verification Steps


1. Clear storage and load the app. 
2. Check network for the graphql for `createCart`. Note the response cartId.
3. Check local storage and you should see the cartId in local storage.
4. Add an item to cart.
5. Update an item in cart.
6. Remove an item from cart.
7. Remove all items from cart.
8. Sign in, add an item, then expire the cart from the admin console and try to remove the item. It should still function same as it did before.

## Screenshots / Screen Captures (if appropriate)

## Checklist

* I have updated the documentation accordingly, if necessary.
* I have added tests to cover my changes, if necessary.
